File: scripts/p11_bruteforce.py
#!/usr/bin/env python3
"""Path B: brute-force enumeration of all 2^(nm) occupancy patterns.

Completely independent of the transfer matrix in p11_square_dp.py: each
configuration is materialised as a grid, clusters are found with BFS on the
4-neighbour graph, and spanning (a cluster touching rows 1 and m) is decided
directly.  Counts A_{n,m}(k) are tallied by number of occupied sites.

Used to validate the transfer-matrix state canonicalization on small widths
(acceptance requirement of Issue 11).  Feasible for nm <= 16 or so in pure
Python (2^16 = 65536 configurations).
"""

from __future__ import annotations

import argparse
import json
import sys


def brute_force_A(n, m):
    nm = n * m
    A = [0] * (nm + 1)
    for mask in range(1 << nm):
        k = bin(mask).count("1")
        occ = [(mask >> (i * n + j)) & 1 for i in range(m) for j in range(n)]
        # BFS over occupied sites; row index i in 0..m-1 corresponds to
        # paper rows 1..m; spanning touches i=0 and i=m-1.
        seen = [False] * nm
        spans = False
        for start in range(nm):
            if not occ[start] or seen[start]:
                continue
            stack = [start]
            seen[start] = True
            touch_top = touch_bottom = False
            while stack:
                v = stack.pop()
                i, j = divmod(v, n)
                if i == 0:
                    touch_top = True
                if i == m - 1:
                    touch_bottom = True
                for di, dj in ((1, 0), (-1, 0), (0, 1), (0, -1)):
                    ii, jj = i + di, j + dj
                    if 0 <= ii < m and 0 <= jj < n:
                        w = ii * n + jj
                        if occ[w] and not seen[w]:
                            seen[w] = True
                            stack.append(w)
            if touch_top and touch_bottom:
                spans = True
                break
        if spans:
            A[k] += 1
    return A


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--n", type=int, required=True)
    ap.add_argument("--m", type=int, default=None)
    args = ap.parse_args()
    m = args.m or args.n
    A = brute_force_A(args.n, m)
    assert A[m] == args.n, f"A[{m}] = {A[m]} != n={args.n}"
    parity = sum(c * (-1) ** k for k, c in enumerate(A) if c)
    print(json.dumps({
        "path": "brute_force",
        "n": args.n, "m": m,
        "A": A,
        "F_minus1": str(parity),
    }))


if __name__ == "__main__":
    main()

File: scripts/p11_bruteforce.cpp
// Path B (fast): brute-force enumeration of all 2^(nm) occupancy patterns on
// the free-boundary n x m grid; union-find spanning check, counts by k.
// Independent of any transfer matrix. Usage: ./p11_bruteforce n [m]
#include <cstdio>
#include <cstdlib>
#include <vector>
using namespace std;

int main(int argc, char** argv) {
    int n = atoi(argv[1]);
    int m = argc > 2 ? atoi(argv[2]) : n;
    int nm = n * m;
    vector<unsigned long long> A(nm + 1, 0);
    vector<int> par(nm + 1), stk;
    for (unsigned long long mask = 0; mask < (1ull << nm); ++mask) {
        int k = __builtin_popcountll(mask & ((1ull << nm) - 1));
        if (nm > 63) { fprintf(stderr, "nm too large\n"); return 1; }
        // union-find with path halving over occupied sites only
        for (int i = 0; i < nm; ++i) par[i] = i;
        auto find = [&](int x) {
            while (par[x] != x) { par[x] = par[par[x]]; x = par[x]; }
            return x;
        };
        for (int i = 0; i < m; ++i)
            for (int j = 0; j < n; ++j) {
                int v = i * n + j;
                if (!((mask >> v) & 1ull)) continue;
                if (j + 1 < n && ((mask >> (v + 1)) & 1ull)) {
                    int a = find(v), b = find(v + 1);
                    if (a != b) par[a] = b;
                }
                if (i + 1 < m && ((mask >> (v + n)) & 1ull)) {
                    int a = find(v), b = find(v + n);
                    if (a != b) par[a] = b;
                }
            }
        bool spans = false;
        for (int j = 0; j < n && !spans; ++j) {
            if (!((mask >> j) & 1ull)) continue;            // top row
            for (int j2 = (m - 1) * n; j2 < m * n; ++j2) {  // bottom row
                if (!((mask >> j2) & 1ull)) continue;
                if (find(j) == find(j2)) { spans = true; break; }
            }
        }
        if (spans) ++A[k];
    }
    printf("{\"path\":\"brute_force_cpp\",\"n\":%d,\"m\":%d,\"A\":[", n, m);
    for (int k = 0; k <= nm; ++k)
        printf("%s%llu", k ? "," : "", A[k]);
    printf("]}\n");
    return 0;
}
